File: apps/api-py/src/lib/instance_settings.py
import json
from typing import Any

from redis.asyncio import Redis
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.models.setting import Setting

CACHE_KEY = "raven:instance_settings"
CACHE_TTL = 60

DEFAULT_SETTINGS: dict[str, Any] = {
    "analytics_retention_days": 90,
    "default_max_tokens": 4096,
    "email_notifications_enabled": False,
    "global_rate_limit_rpm": 60,
    "global_rate_limit_rpd": 1000,
    "instance_name": "Raven",
    "log_request_bodies": False,
    "log_response_bodies": False,
    "max_request_body_size_mb": 10,
    "password_min_length": 8,
    "request_timeout_seconds": 30,
    "resend_api_key": "",
    "resend_from_email": "",
    "session_timeout_hours": 720,
    "signup_enabled": True,
    "webhook_retry_count": 3,
    "webhook_timeout_seconds": 10,
}

PUBLIC_SETTING_KEYS = {"instance_name", "signup_enabled"}


def _cast_value(key: str, raw: str) -> Any:
    default = DEFAULT_SETTINGS.get(key)
    if isinstance(default, bool):
        return raw.lower() in ("true", "1", "yes")
    if isinstance(default, int):
        try:
            return int(raw)
        except ValueError:
            return default
    if isinstance(default, float):
        try:
            return float(raw)
        except ValueError:
            return default
    return raw


async def get_instance_settings(session: AsyncSession, redis: Redis) -> dict[str, Any]:
    cached = await redis.get(CACHE_KEY)
    if cached:
        return json.loads(cached)

    result = await session.execute(select(Setting))
    rows = result.scalars().all()

    merged = dict(DEFAULT_SETTINGS)
    for row in rows:
        merged[row.key] = _cast_value(row.key, row.value)

    await redis.set(CACHE_KEY, json.dumps(merged), ex=CACHE_TTL)
    return merged


async def get_public_settings(session: AsyncSession, redis: Redis) -> dict[str, Any]:
    all_settings = await get_instance_settings(session, redis)
    return {k: v for k, v in all_settings.items() if k in PUBLIC_SETTING_KEYS}


async def invalidate_settings_cache(redis: Redis) -> None:
    await redis.delete(CACHE_KEY)

File: apps/api/src/lib/instance-settings.ts
import type { Database } from "@raven/db";
import { settings } from "@raven/db";
import type { Redis } from "ioredis";

export interface InstanceSettings {
  readonly instance_name: string;
  readonly instance_url: string;
  readonly analytics_retention_days: number;
  readonly signup_enabled: boolean;
  readonly session_timeout_hours: number;
  readonly password_min_length: number;
  readonly global_rate_limit_rpm: number;
  readonly global_rate_limit_rpd: number;
  readonly max_request_body_size_gb: number;
  readonly request_timeout_seconds: number;
  readonly default_max_tokens: number;
  readonly log_request_bodies: boolean;
  readonly log_response_bodies: boolean;
  readonly webhook_timeout_seconds: number;
  readonly webhook_retry_count: number;
  readonly email_notifications_enabled: boolean;
  readonly knowledge_enabled: boolean;
  readonly bigrag_url: string;
  readonly bigrag_api_key: string;
  readonly notify_on_budget_exceeded: boolean;
  readonly notify_on_provider_error_spike: boolean;
}

const DEFAULTS: Record<string, string> = {
  analytics_retention_days: "365",
  bigrag_api_key: "",
  bigrag_url: "",
  default_max_tokens: "4096",
  email_notifications_enabled: "false",
  global_rate_limit_rpd: "1000",
  global_rate_limit_rpm: "60",
  instance_name: "Raven",
  instance_url: "",
  knowledge_enabled: "false",
  log_request_bodies: "true",
  log_response_bodies: "false",
  max_request_body_size_gb: "1",
  notify_on_budget_exceeded: "true",
  notify_on_provider_error_spike: "true",
  password_min_length: "8",
  request_timeout_seconds: "300",
  session_timeout_hours: "24",
  signup_enabled: "true",
  webhook_retry_count: "3",
  webhook_timeout_seconds: "10"
};

const CACHE_KEY = "instance:settings";

const toBool = (v: string | undefined): boolean => v === "true";
const toNum = (v: string | undefined, fallback: number): number => {
  const n = Number(v);
  return Number.isFinite(n) ? n : fallback;
};

const parse = (raw: Record<string, string>): InstanceSettings => ({
  analytics_retention_days: toNum(raw.analytics_retention_days, 365),
  bigrag_api_key: raw.bigrag_api_key || "",
  bigrag_url: raw.bigrag_url || "",
  default_max_tokens: toNum(raw.default_max_tokens, 4096),
  email_notifications_enabled: toBool(raw.email_notifications_enabled),
  global_rate_limit_rpd: toNum(raw.global_rate_limit_rpd, 1000),
  global_rate_limit_rpm: toNum(raw.global_rate_limit_rpm, 60),
  instance_name: raw.instance_name || "Raven",
  instance_url: raw.instance_url || "",
  knowledge_enabled: toBool(raw.knowledge_enabled),
  log_request_bodies: toBool(raw.log_request_bodies),
  log_response_bodies: toBool(raw.log_response_bodies),
  max_request_body_size_gb: toNum(raw.max_request_body_size_gb, 1),
  notify_on_budget_exceeded: toBool(raw.notify_on_budget_exceeded),
  notify_on_provider_error_spike: toBool(raw.notify_on_provider_error_spike),
  password_min_length: toNum(raw.password_min_length, 8),
  request_timeout_seconds: toNum(raw.request_timeout_seconds, 300),
  session_timeout_hours: toNum(raw.session_timeout_hours, 24),
  signup_enabled: toBool(raw.signup_enabled),
  webhook_retry_count: toNum(raw.webhook_retry_count, 3),
  webhook_timeout_seconds: toNum(raw.webhook_timeout_seconds, 10)
});

export const getInstanceSettings = async (
  db: Database,
  redis?: Redis
): Promise<InstanceSettings> => {
  if (redis) {
    const cached = await redis.get(CACHE_KEY).catch(() => null);
    if (cached) return JSON.parse(cached) as InstanceSettings;
  }

  const rows = await db.select().from(settings);
  const raw = { ...DEFAULTS };
  for (const row of rows) {
    if (row.key in DEFAULTS) raw[row.key] = row.value;
  }

  const parsed = parse(raw);

  if (redis) {
    await redis.set(CACHE_KEY, JSON.stringify(parsed)).catch(() => undefined);
  }

  return parsed;
};

/** Invalidate the cached settings (call after admin updates settings) */
export const invalidateSettingsCache = async (redis: Redis): Promise<void> => {
  await redis.del(CACHE_KEY).catch(() => undefined);
};
